contracts/true-currencies/TrueCurrencyWithPoR.sol

```diff
@@ -0,0 +1,99 @@
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+/**
+ * @title TrueCurrencyWithPoR
+ * @dev TrueCurrencyPoR is an ERC20 with blacklist & redemption addresses.
+ *  Please see TrueCurrency for the implementation that this contract inherits from.
+ *  This contract implements an additional check against a Proof-of-Reserves feed before
+ *  allowing tokens to be minted.
+ */
+abstract contract TrueCurrencyWithPoR is TrueCurrency, IPoRToken {
+    using SafeMath for uint256;
+
+    uint256 public constant MAX_AGE = 7 days;
+
+    constructor() public {
+        chainReserveHeartbeat = MAX_AGE;
+    }
+
+    /**
+     * @notice Overriden mint function that checks the specified proof-of-reserves feed to
+     * ensure that the total supply of this TrueCurrency is not greater than the reported
+     * reserves.
+     * @dev The proof-of-reserves check is bypassed if feed is not set.
+     * @param account The address to mint tokens to
+     * @param amount The amount of tokens to mint
+     */
+    function _mint(address account, uint256 amount) internal virtual override {
+        if (chainReserveFeed == address(0)) {
+            super._mint(account, amount);
+            return;
+        }
+
+        // Get latest proof-of-reserves from the feed
+        (, int256 answer, , uint256 updatedAt, ) = IChainlinkAggregatorV3(chainReserveFeed).latestRoundData();
+        require(answer > 0, "TrueCurrency: Invalid answer from PoR feed");
+
+        // Check the answer is fresh enough (i.e., within the specified heartbeat)
+        uint256 oldestAllowed = block.timestamp.sub(chainReserveHeartbeat, "TrueCurrency: Invalid timestamp from PoR feed");
+        require(updatedAt >= oldestAllowed, "TrueCurrency: PoR answer too old");
+
+        uint256 reserves = uint256(answer);
+
+        // Get required info about total supply & decimals
+        uint256 currentSupply = totalSupply();
+        uint8 trueDecimals = decimals();
+        uint8 reserveDecimals = IChainlinkAggregatorV3(chainReserveFeed).decimals();
+        // Normalise TrueCurrency & reserve decimals
+        if (trueDecimals < reserveDecimals) {
+            currentSupply = currentSupply.mul(10**uint256(reserveDecimals - trueDecimals));
+        } else if (trueDecimals > reserveDecimals) {
+            reserves = reserves.mul(10**uint256(trueDecimals - reserveDecimals));
+        }
+
+        // Check that after minting more tokens, the total supply would NOT exceed the reserves
+        // reported by the latest valid proof-of-reserves feed.
+        require(currentSupply + amount <= reserves, "TrueCurrency: total supply would exceed reserves after mint");
+        super._mint(account, amount);
+    }
+
+    /**
+     * @notice Sets a new feed address
+     * @dev Admin function to set a new feed
+     * @param newFeed Address of the new feed
+     */
+    function setChainReserveFeed(address newFeed) external override onlyOwner returns (uint256) {
```

Comment:
I would recommend against a zero address check, since `address(0)` in this current implementation acts as a sentinel value that indicates proof of reserve is disabled.

This might be necessary to call in case, for instance, the Chainlink aggregator goes down for a period that is longer than `MAX_AGE`.